### PR TITLE
Add B-frame decoding support for Main and High Profile.

### PR DESCRIPTION
**Pull request by xiaotiansf**

### CODE REVIEW

**Comment by sijchen** [comment]
@xiaotiansf Thank you for the effort! I can see this is a very big PR with only one commit. This is not what we commonly expect since the code review will be not that easy, but we will try our best. Please bear us some time for the code review. Meanwhile if you are still willing to contribute, we will suggest adding some test cases for B frames.

---

@GuangweiWang / @huili2 , could you please kindly review the diff in following files, thx: 
- [ ] codec/vsproject/windecoder.sln
- [ ] codec/vsproject/windecoder.vcxproj
- [ ] codec/vsproject/windecoder.vcxproj.filters

**Comment by xiaotiansf** [comment]
Sijia,

Thank you so much and I'll add some test cases.

-Xiaotian

File: codec/decoder/core/src/pic_queue.cpp [hunk 1]
```diff
@@ -118,6 +131,21 @@ void FreePicture (PPicture pPic, CMemoryAlign* pMa) {
       pMa->WelsFree (pPic->pBuffer[0], "pPic->pBuffer[0]");
     }
 
+		if (pPic->pMbType) {
```
Comment (ruil2) [inline codec/decoder/core/src/pic_queue.cpp]: please use astyle to format source code

Comment (xiaotianshicomcast) [reply codec/decoder/core/src/pic_queue.cpp]: You mean to make the format looking like the following?
How to use "astyle" tool?
Thanks.
void FreePicture (PPicture pPic, CMemoryAlign* pMa) {
  if (NULL != pPic) {
    if (pPic->pBuffer[0]) {
      pMa->WelsFree (pPic->pBuffer[0], "pPic->pBuffer[0]");
      pPic->pBuffer[0] = NULL;
    }

    if (pPic->pMbType) {
      pMa->WelsFree(pPic->pMbType, "pPic->pMbType");
      pPic->pMbType = NULL;
    }
		
    for (int32_t listIdx = LIST_0; listIdx < LIST_A; ++listIdx) {
      if (pPic->pMv[listIdx]) {
        pMa->WelsFree(pPic->pMv[listIdx], "pPic->pMv[]");
        pPic->pMv[listIdx] = NULL;
     }
			
     if (pPic->pRefIndex[listIdx]) {
       pMa->WelsFree(pPic->pRefIndex[listIdx], "pPic->pRefIndex[]");
       pPic->pRefIndex[listIdx] = NULL;
     }
   }
   pMa->WelsFree (pPic, "pPic");
   pPic = NULL;
 }
}

Comment (ruil2) [reply codec/decoder/core/src/pic_queue.cpp]: code style is defined in build/astyle.cfg, you can type "astyle --options=./build/astyle.cfg *.cpp or *h"  to format the code. thanks!

Comment (xiaotianshicomcast) [reply codec/decoder/core/src/pic_queue.cpp]: I see. Thx.

**Comment by xiaotianshicomcast** [comment]
Sijia,
The usage is mv_pred.cpp 
subMbType = SUB_MB_TYPE_8x8 | MB_TYPE_P0L0 | MB_TYPE_P0L1 | MB_TYPE_DIRECT
I'll see if there is any conflict. Visually inspection shows that it is possible to change back to 0x00000001.  I can definitely try.

**Comment by GuangweiWang** [comment]
@xiaotiansf 

I encountered an error while compiling your code on Mac OSX for Android platform:

make OS=android ARCH=arm NDKROOT=$ANDROID_NDK TARGET=android-19

/Users/jenkins/Library/Android/android-ndk-r13b/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-g++ -O3 -DNDEBUG -DHAVE_NEON -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -DANDROID_NDK -fpic --sysroot=/Users/jenkins/Library/Android/android-ndk-r13b/platforms/android-12/arch-arm -MMD -MP -DGENERATED_VERSION_HEADER -fno-rtti -fno-exceptions -I./codec/api/svc -I./codec/common/inc -Icodec/common/inc -Dandroid_getCpuIdArm=wels_getCpuIdArm -Dandroid_setCpuArm=wels_setCpuArm -Dandroid_getCpuCount=wels_getCpuCount -Dandroid_getCpuFamily=wels_getCpuFamily -Dandroid_getCpuFeatures=wels_getCpuFeatures -Dandroid_setCpu=wels_setCpu  -I/Users/jenkins/Library/Android/android-ndk-r13b/sources/android/cpufeatures -c -o codec/common/src/common_tables.o codec/common/src/common_tables.cpp
In file included from ./codec/common/inc/wels_common_defs.h:37:0,
                 from codec/common/src/common_tables.cpp:33:
./codec/common/inc/typedefs.h:39:19: fatal error: cstddef: No such file or directory
 #include <cstddef>
                   ^
compilation terminated.
make: *** [codec/common/src/common_tables.o] Error 1


could you help to check this issue? Thanks.

**Comment by GuangweiWang** [comment]
@xiaotiansf 
Thanks for the quick fix. The code can be built successfully for android platform now.

**Comment by sijchen** [comment]
@xiaotiansf unfortunately your latest commit is introducing failed CI history again, in such a case we cannot merge this PR. Would you please make another PR and keep the CI history clean? Thank you!

**Comment by sijchen** [comment]
@xiaotiansf  the latest status of the mentioned PR has again been polluted with some failed CI history, after the great attempt of removing them previously. We would like to have the failed CI history removed, before we can merge the PR. Thanks.

**Comment by sijchen** [comment]
the code diff generally LGTM, btw.

**Comment by GuangweiWang** [comment]
@xiaotiansf 

I pull your latest code for undefined behavior sanitizer testing and got some runtime errors like this:

codec/decoder/core/src/mv_pred.cpp:846:32: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:41: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:61: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:81: runtime error: left shift of negative value -1
codec/decoder/core/src/mv_pred.cpp:907:32: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:124:41: runtime error: left shift of negative value -1
codec/decoder/core/src/mv_pred.cpp:937:32: runtime error: left shift of negative value -1
codec/decoder/core/src/mv_pred.cpp:846:32: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:41: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:61: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:81: runtime error: left shift of negative value -1
codec/decoder/core/src/mv_pred.cpp:907:32: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:124:41: runtime error: left shift of negative value -1
codec/decoder/core/src/mv_pred.cpp:937:32: runtime error: left shift of negative value -1
codec/decoder/core/src/rec_mb.cpp:222:13: runtime error: index -1 out of bounds for type 'PPicture [17]'
codec/decoder/core/src/rec_mb.cpp:224:40: runtime error: member access within null pointer of type 'struct SPicture'
codec/decoder/core/src/mv_pred.cpp:846:32: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:124:41: runtime error: left shift of negative value -1
codec/decoder/core/src/mv_pred.cpp:937:32: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:41: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:61: runtime error: left shift of negative value -1
codec/decoder/core/src/parse_mb_syn_cabac.cpp:107:81: runtime error: left shift of negative value -1
codec/decoder/core/src/mv_pred.cpp:907:32: runtime error: left shift of negative value -1


could you help to check this? thanks.

**Comment by xiaotianshi2** [comment]
@GuangweiWang,

Thanks for testing. Can you tell me the tools to run sanitizer testing?

**Comment by xiaotianshi2** [comment]
@GuangweiWang,

I think I understand:
in file codec/decoder/core/src/mv_pred.cpp:846:32: runtime error: left shift of negative value -1 
845 void UpdateP16x16MotionInfo (PDqLayer pCurDqLayer, int32_t listIdx, int8_t iRef, int16_t iMVs[2]) {
846  const int16_t kiRef2 = (iRef << 8) | iRef;

According to  The iRef can be the following value:

#define REF_NOT_IN_LIST  -1  //intra

My question is that:

Does source code "https://github.com/cisco/openh264" have the same issue?

Thanks.

**Comment by xiaotianshi2** [comment]
@GuangweiWang,

Probably I can change the following defines


#define REF_NOT_IN_LIST  -1  //intra

as 

#define REF_NOT_IN_LIST  0xFF  //intra

if that really needs to be fixed.

**Comment by GuangweiWang** [comment]
@xiaotiansf 

I use the Clang undefined behavior detector to do that testing. attach my test step:
1. merge your pr#2983 to master.
2. export : CC=clang, CXX=clang++, CFLAGS+= -fsanitize=undefined -fno-sanitize=vptr, LDFLAGS+=-fsanitize=undefined -fno-sanitize=vptr
3. make gtest-bootstrap
4. make clean
5. DEBUGSYMBOLS=True
6. run ./codec_unittest
7. take some *.264 bitstream files for decoder test.

**Comment by GuangweiWang** [comment]
@xiaotiansf 

source code "https://github.com/cisco/openh264" dose not have the same issue.

**Comment by xiaotianshi2** [comment]
@GuangweiWang,

Thank you for your information. That's very helpful.

**Comment by xiaotianshi2** [comment]
@GuangweiWang,

Adding CFLAGS+= -fsanitize=undefined -fno-sanitize=vptr causes linking error It causes the
clang: error: unsupported argument CFLAGS+= -fsanitize=undefined -fno-sanitize=vptr
Do you know how to fix the clang error?

Also I have checked in a fix to address the  issue of undefined behavior sanitizer. Could you run test again?

Thanks a lot.

**Comment by xiaotiansf** [comment]
@GuangweiWang,

I am able to fix the clang error:

Here is what the steps which I take on OSX 10.13:

1. Get the latest source code including my today's checkins which are intended to fix your reported issue.
2. Edit Makefile to include 

DEBUGSYMBOLS=True
CFLAGS += -fsanitize=address
CFLAGS += -fsanitize=undefined
CFLAGS += -fno-sanitize=vptr
LDFLAGS += -fsanitize=address
LDFLAGS += -fsanitize=undefined
LDFLAGS += -fno-sanitize=vptr

3. make gtest-bootstrap
4. make clean
5. ./run_Test.sh unittest

The result shows no runtime error anymore!
Can you verify my result?

Thank you.

**Comment by GuangweiWang** [comment]
@xiaotiansf 
ok, I will.

**Comment by xiaotiansf** [comment]
@GuangweiWang,

If you still reproduce the issue, could you give me much detailed step-by-step instructions including scripts and sample assets you used?  I try to wrap things up by tomorrow.

Thank you so much.

**Comment by GuangweiWang** [comment]
@xiaotiansf 

that issue seems to be fixed. thanks!

**Pull request closed without merging**